### PR TITLE
Change SLA Management / add OLA object

### PR DESCRIPTION
**Pull request by amma35**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | 
| Fixed tickets | #1911

### CODE REVIEW

**Comment by flegastelois** [comment]
**This review is not a technical review.**

On fresh GLPI install, a first functional test on Ticket form looks good to me  : 
- [x] Create SLM : OK
- [x] Add 2 SLA (TTO/TTR) into SLM : OK
- [x] Add 4 OLA (TTO/TTR for 2 internal groups) into SLM : OK
- [x] Add business rules with "Technician Group" Criterias and "OLAs" Actions : OK
- [x] Add business rules with "Type, Category" Criterias and "affect SLAs, Technician Group" Actions : OK
- [x] Add ticket with specific "Type, Category" : SLA OK, GROUP OK, OLA OK
- [x] Escalation to another group on ticket form : new OLA OK

Need the same tests on Problem/Change.
Need tests for Cron/Notification.
Need tests with data Migration from and under GLPI version.
Need doc improvement.

This is a good start @amma35  !

File: front/olalevel.form.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
```
Comment (trasher) [inline front/olalevel.form.php]: Please precise version added

File: front/olalevel.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
+* @since version 9.1
```
Comment (trasher) [inline front/olalevel.php]: 9.2

File: front/olalevelaction.form.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
+* @since version 0.85
```
Comment (trasher) [inline front/olalevelaction.form.php]: 9.2

File: front/olalevelcriteria.form.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
+* @since version 0.85
```
Comment (trasher) [inline front/olalevelcriteria.form.php]: 9.2

File: front/slm.form.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
```
Comment (trasher) [inline front/slm.form.php]: Version added

File: inc/ola.class.php [hunk 6]
```diff
@@ -35,34 +35,29 @@
 }
 
 /**
- * SLT Class
+ * OLA Class
  * @since version 9.1
```
Comment (trasher) [inline inc/ola.class.php]: 9.2

File: inc/olalevel.class.php [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
```
Comment (trasher) [inline inc/olalevel.class.php]: Version added

File: inc/olalevel_ticket.class.php [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+/** @file
+* @brief
+*/
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+/// Class OLALevel
```
Comment (trasher) [inline inc/olalevel_ticket.class.php]: version added

File: inc/ticket.class.php [hunk 9]
```diff
@@ -332,46 +332,84 @@ function canMassiveAction($action, $field, $value) {
...
 
+   /**
+    * Get Datas to be added for OLA add
+    *
+    * @param $olas_id      OLA id
+    * @param $entities_id  entity ID of the ticket
+    * @param $date         begin date of the ticket
+    * @param $type         type of OLA
+    *
+    * @since version 9.1 (before getDatasToAddOla without type parameter)
```
Comment (trasher) [inline inc/ticket.class.php]: 9.2 probably

File: front/slm.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
```
Comment (trasher) [inline front/slm.php]: Missing version added

File: inc/olalevelaction.class.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
```
Comment (trasher) [inline inc/olalevelaction.class.php]: Missing version added

File: inc/olalevelcriteria.class.php [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+* @brief
+* @since version 0.84
```
Comment (trasher) [inline inc/olalevelcriteria.class.php]: Wrong version

File: inc/sla.class.php [hunk 13]
```diff
@@ -40,23 +40,29 @@
 
 /**
  * SLA Class
+ * @since version 9.1
```
Comment (trasher) [inline inc/sla.class.php]: Wrong version

File: inc/sla.class.php [hunk 14]
```diff
@@ -71,18 +77,77 @@ function defineTabs($options = []) {
...
-      $this->addStandardTab('SLT', $ong, $options);
-      $this->addStandardTab('Log', $ong, $options);
+      $this->addStandardTab('SlaLevel', $ong, $options);
+      $this->addStandardTab('Rule', $ong, $options);
+      $this->addStandardTab('Ticket', $ong, $options);
 
       return $ong;
    }
 
+   /**
```
Comment (trasher) [inline inc/sla.class.php]: Useless, please drop.

File: inc/sla.class.php [hunk 15]
```diff
@@ -71,18 +77,77 @@ function defineTabs($options = []) {
...
+      if ($slm->getFromDB($this->fields['slms_id'])) {
+         $this->fields['calendars_id'] = $slm->fields['calendars_id'];
+         return true;
+      } else {
+         return false;
+      }
+   }
+
+
+   /**
```
Comment (trasher) [inline inc/sla.class.php]: Useless, please drop

File: inc/sla.class.php [hunk 16]
```diff
@@ -108,20 +192,430 @@ function showForm($ID, $options = []) {
...
+            Html::closeForm();
+         }
+      } else {
+         echo __('No item to display');
+      }
+      echo "</div>";
+   }
+
+
+   /**
```
Comment (trasher) [inline inc/sla.class.php]: Useless, please drop

File: inc/sla.class.php [hunk 17]
```diff
@@ -108,20 +192,430 @@ function showForm($ID, $options = []) {
...
+               }
+               return self::createTabEntry(self::getTypeName(Session::getPluralNumber()), $nb);
+         }
+      }
+      return '';
+   }
+
+
+   /**
+    *
```
Comment (trasher) [inline inc/sla.class.php]: Missing description, missing parameters types

Comment (trasher) [inline inc/sla.class.php]: BTW, if this method is inherited; you can drop the phpdoc here.

File: inc/sla.class.php [hunk 18]
```diff
@@ -108,20 +192,430 @@ function showForm($ID, $options = []) {
...
+            break;
+      }
+      return true;
+   }
+
+
+   /**
+    * Get SLA data by type and ticket
+    *
+    * @param $tickets_id
```
Comment (trasher) [inline inc/sla.class.php]: Missing types and params descriptions

Comment (trasher) [inline inc/sla.class.php]: Also missing @return

File: inc/sla.class.php [hunk 19]
```diff
@@ -108,20 +192,430 @@ function showForm($ID, $options = []) {
...
+            break;
+      }
+      return $this->getFromDBByQuery("INNER JOIN `glpi_tickets` ON (`glpi_tickets`.`$field` = `".$this->getTable()."`.`id`) WHERE `glpi_tickets`.`id` = '".$tickets_id."' LIMIT 1");
+   }
+
+
+    /**
+    * Get SLA datas by condition
+    *
+    * @param $condition
```
Comment (trasher) [inline inc/sla.class.php]: Missing type, param description and @return

File: inc/sla.class.php [hunk 20]
```diff
@@ -108,20 +192,430 @@ function showForm($ID, $options = []) {
...
+   **/
+   function getSlaData($condition) {
+      return $this->find($condition);
+   }
+
+
+   /**
+    * Get SLA table fields
+    *
+    * @param $type
```
Comment (trasher) [inline inc/sla.class.php]: Missing type & description

File: inc/sla.class.php [hunk 21]
```diff
@@ -108,20 +192,430 @@ function showForm($ID, $options = []) {
...
+            break;
+      }
+      return [$dateField, $slaField];
+   }
+
+
+   /**
+    * Show SLA for ticket
+    *
+    * @param $ticket      Ticket item
```
Comment (trasher) [inline inc/sla.class.php]: Missing tyes, descriptions, @return

File: inc/sla.class.php [hunk 22]
```diff
@@ -149,43 +643,357 @@ function getSearchOptionsNew() {
...
+         'name'               => __('Comments'),
+         'datatype'           => 'text'
+      ];
+
       return $tab;
    }
 
+
+   /**
+    * @param $field
```
Comment (trasher) [inline inc/sla.class.php]: Missing method description, params types, params descriptions, @return

Comment (trasher) [inline inc/sla.class.php]: Ok well.. Same for all other methods :)

File: inc/sla.class.php [hunk 23]
```diff
@@ -149,43 +643,357 @@ function getSearchOptionsNew() {
...
+         }
+      }
+      return 0;
+   }
+
+
+   /**
+    * Get execution date of a sla level
+    *
+    * @param $start_date         datetime    start date
```
Comment (trasher) [inline inc/sla.class.php]: Type first

File: inc/sla.class.php [hunk 24]
```diff
@@ -149,43 +643,357 @@ function getSearchOptionsNew() {
...
+    * @param $slalevels_id       integer     sla level id
+    * @param $additional_delay   integer     additional delay to add or substract (for waiting time)
+    *                                        (default 0)
+    *
+    * @return execution date time (null if sla not exists)
```
Comment (trasher) [inline inc/sla.class.php]: missing return type

File: inc/slm.class.php [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+/** @file
+* @brief
+*/
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+/**
+ * SLM Class
```
Comment (trasher) [inline inc/slm.class.php]: missing @since

File: install/update_91_92.php [hunk 26]
```diff
@@ -648,6 +648,180 @@ function update91to92() {
...
+                    `date` datetime DEFAULT NULL,
+                    PRIMARY KEY (`id`),
+                    KEY `tickets_id` (`tickets_id`),
+                    KEY `olalevels_id` (`olalevels_id`),
+                    KEY `unicity` (`tickets_id`,`olalevels_id`)
+                  ) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;";
+         $DB->queryOrDie($query, "9.2 add table glpi_olalevels_tickets");
+      }
+
+      $query = "INSERT INTO `glpi_crontasks`
```
Comment (trasher) [inline install/update_91_92.php]: Use cron register method?

File: install/update_91_92.php [hunk 27]
```diff
@@ -648,6 +648,180 @@ function update91to92() {
...
+                       (`itemtype`, `name`, `frequency`, `param`, `state`, `mode`, `allowmode`,
+                        `hourmin`, `hourmax`, `logs_lifetime`, `lastrun`, `lastcode`, `comment`)
+                VALUES ('OlaLevel_Ticket', 'olaticket', 604800, NULL, 0, 1, 3,
+                        0, 24, 10, NULL, NULL, NULL); ";
+      $DB->queryOrDie($query, "9.2 populate glpi_crontasks for olaticket");
+
+      // Changing the structure of the table 'glpi_slas'
+      $migration->renameTable('glpi_slas', 'glpi_slms');
```
Comment (trasher) [inline install/update_91_92.php]: For this one and above: must be checked to be re-entrant

**Comment by orthagh** [comment]
Some functional comments:

- `Administration > SLM` should be renamed to `Administration > Service levels` (acronyms is bad)
- in administration sla/ola forms, we could keep theses acronyms, but an explicative text (above the add buttons) could be a great enhancement.
- not mandatory for merge, but in ticket form, the 4 field for service levels take a lot of space.
   possible evolutions:
   - display internal fields in the same row as normal ones (with a little indication for differentiation)
   - more, display them only if an slm with an existing ola is defined, theses fields are not needed in other cases, i think
   - or, completely different, create a sla tab, which displays internal dates (normal dates can be kept in main form but repeated in this new tab) and escalation levels timeline 
- non blocking for merge, but for 9.2 release: contribute a [user doc](https://github.com/glpi-project/doc) enhancement
- i found a notice (maybe my fault with recent merge):
```
PHP Notice: Undefined index: type in .../inc/slalevel.class.php at line 391
```
- it seem i can't affect manually a sla to a ticket, it loads indefinitely:
image

File: install/update_91_92.php (hunk 26 above)
Comment (amma35) [inline install/update_91_92.php]: The method is present in the file olalevel_ticket.class.php

**Comment by amma35** [comment]
For the notice, it's linked to commit : https://github.com/glpi-project/glpi/commit/9883e6166bd220b2aa8888f99dfa6109049c686a

File: install/update_91_92.php (hunk 26 above)
Comment (trasher) [inline install/update_91_92.php]: What I told you was to use the `Crontask::register` method that will automatically check if the task already exists before adding it.
Currently; this method is working only for plugins... I've changed that in one of my recent PR; that has not yet been merged (will be today tought -- see https://github.com/glpi-project/glpi/pull/2438/files#diff-1454c423049813362d0918fb1f2dc346R1387)

**Comment by trasher** [comment]
Thank you very much for the work on the phpdoc; should be OK now :)

As of the update file... You must check table/field/index presence on every add/drop/rename/etc. Same update must be able to run several times on the same database.
As an exemple, if you rename `glpi_a` to `glpi_b` and then run again the update script; you'll get an SQL error because o`glpi_a` no longer exists...

You can make sure the script can be run severa times using a command like `php tools/cliupdate.php --dev --force`

**Comment by orthagh** [comment]
Ok for at this point (and after some additional fixes).
We merge, and we can discuss for enhancement (but atm, the work is sufficient)

**Pull request merged**